sql/python/sqlflow_submitter/xgboost/sqlflow_data_source.py

```diff
@@ -0,0 +1,207 @@
+    is_tf_integrated: bool
+    # FeatureMetas are useless temporarily, since tf transformation is not supported.
+    x: typing.List[typing.Dict]
+    label: typing.Dict = None
+    group: typing.Dict = None
+    weight: typing.Dict = None
+    output_table: str = None
+    write_batch_size: int = 1024
+
+
+class SqlflowDataSource(DataSource):
+    def __init__(self, rank: int, num_worker: int,
+                 column_conf: config_fields.ColumnFields,
+                 source_conf):
+        super().__init__(rank, num_worker, column_conf, source_conf)
+        if not isinstance(source_conf, SqlflowDSConfig):
+            raise XGBoostError("SqlflowDataSource: invalid source conf")
+
+        # TODO: support tf.feature_column transformation
+        if source_conf.is_tf_integrated:
+            raise XGBoostError('So far, tf transformation is not supported in xgboost job.')
+
+        self._train = source_conf.is_train
+        self._rcd_builder = RecordBuilder(column_conf.features)
+
+        spec_temp = {'is_sparse': False, 'shape': [1], 'delimiter': ''}
+        metas = {}
+        col_names = []
+        for col in column_conf.features.columns:
+            col_names.append(col)
+            metas[col] = spec_temp
+        self._feature_len = len(col_names)
+
+        self._label_offset = -1
+        self._group_offset = -1
+        self._weight_offset = -1
+        self._append_indices = []
+        if self._train:
+            # label, group, weight column are training specialized
+            if column_conf.label:
+                self._label_offset = len(col_names)
+                col_names.append(column_conf.label)
+                metas[column_conf.label] = spec_temp
+
+            if column_conf.group:
+                self._group_offset = len(col_names)
+                col_names.append(column_conf.group)
+                metas[column_conf.group] = spec_temp
+
+            if column_conf.weight:
+                self._weight_offset = len(col_names)
+                col_names.append(column_conf.weight)
+                metas[column_conf.weight] = spec_temp
+        else:
+            # append columns are prediction specialized
+            for col in column_conf.append_columns:
+                if col in metas:
+                    self._append_indices.append(col_names.index(col))
+                else:
+                    self._append_indices.append(len(col_names))
+                    col_names.append(col)
+                    metas[col] = spec_temp
+
+            self._result_schema = {'append_columns': column_conf.append_columns or []}
+            self._result_schema.update(column_conf.result_columns._asdict())
+
+        conn = connect(**source_conf.db_config)
+
+        self._reader = db_generator(
+            driver=source_conf.db_config['driver'],
+            conn=conn,
+            statement=source_conf.standard_select,
+            feature_column_names=col_names,
+            label_column_name=None,
```

Comment:
Just confusion why the label column is None.

Reply:
Because all column fields in use have already been included in feature_column_names.

https://github.com/sql-machine-learning/sqlflow/pull/647/files#diff-10bbf40d8866f298982cb34c597430d9R73